### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by REASY**
## Main changes in PR:
- Replace data structures to appropriate one to reduce runtime
- Explicitly put check for the log level because Actor Logging need to calculate args before calling logging method
- Intensive operation to get idle vehicles is cached and passed instead of re-computing it

## Run params:
- Config: `production/sfbay/smart/smart-b-hightech-2040.conf`, for 5 iterations
- Baseline: branch [art/#2223-prod-dev-updated](https://github.com/LBNL-UCB-STI/beam/tree/art/%232223-prod-dev-updated), commit 674b838aefac6c9ed16de6bfa3fd17999b8235e2, result: https://s3.us-east-2.amazonaws.com/beam-outputs/index.html#output/sfbay/sfbay-smart-b-ht-2040__2019-10-25_19-37-05
- Optimized: branch [art/#2223-prod-dev-updated](https://github.com/LBNL-UCB-STI/beam/tree/art/%232223-prod-dev-updated), commit 9c0cb1fe62b2a1783aded2fdeddba493c820971a, result: https://s3.us-east-2.amazonaws.com/beam-outputs/index.html#output/sfbay/sfbay-smart-b-ht-2040__2019-10-27_09-01-07/

## Results
Run | Iter 0, s | Iter 1, s | Iter 2, s | Iter 3, s | Iter 4, s | Iter 5, s | Sum, s
-- | -- | -- | -- | -- | -- | -- | --
Baseline | 4212 | 3145 | 3203 | 3102 | 3253 | 3133 | 20048
Optimized | 2775 | 1927 | 1886 | 1895 | 1901 | 2005 | 12389
Baseline /   Optimized | 1.52 | 1.63 | 1.70 | 1.64 | 1.71 | 1.56 | 1.62

### Baseline overall
image
### Baseline 1-th iteration
image

### Optimized overall
image
### Optimized 1-th iteration
image

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2242)

### CODE REVIEW

File: src/main/scala/beam/agentsim/agents/ridehail/VehicleCentricMatchingForRideHail.scala [hunk 1]
```diff
@@ -150,15 +150,15 @@ class VehicleCentricMatchingForRideHail(
   }
 
   private def greedyAssignment(trips: List[AssignmentKey]): List[AssignmentKey] = {
-    val Rok = mutable.ListBuffer.empty[CustomerRequest]
-    val Vok = mutable.ListBuffer.empty[VehicleAndSchedule]
+    val Rok = mutable.HashSet.empty[CustomerRequest]
```
Comment (REASY) [inline src/main/scala/beam/agentsim/agents/ridehail/VehicleCentricMatchingForRideHail.scala]: @haitamlaarabi please, take a look to these changes. It should not break anything, but still.

File: src/main/scala/beam/agentsim/agents/ridehail/AlonsoMoraPoolingAlgForRideHail.scala [hunk 2]
```diff
@@ -177,8 +177,8 @@ object AlonsoMoraPoolingAlgForRideHail {
     solutionSpaceSizePerVehicle: Int
   ): List[(RideHailTrip, VehicleAndSchedule, Double)] = {
     import scala.collection.mutable.{ListBuffer => MListBuffer}
-    val Rok = MListBuffer.empty[CustomerRequest]
-    val Vok = MListBuffer.empty[VehicleAndSchedule]
+    val Rok = collection.mutable.HashSet.empty[CustomerRequest]
```
Comment (REASY) [inline src/main/scala/beam/agentsim/agents/ridehail/AlonsoMoraPoolingAlgForRideHail.scala]: @haitamlaarabi please, take a look to these changes. It should not break anything, but still.

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala [hunk 3]
```diff
@@ -1364,13 +1364,15 @@ class RideHailManager(
...
     finalTriggersToSchedule: Vector[ScheduleTrigger]
   ): Unit = {
-    log.debug(
-      "Removing request: {} pendingAcks: {} pendingRoutes: {} requestBufferSize: {}",
-      requestId,
-      s"(${pendingModifyPassengerScheduleAcks.size}) ${pendingModifyPassengerScheduleAcks.keySet.map(_.toString).mkString(",")}",
-      numPendingRoutingRequestsForReservations,
-      rideHailResourceAllocationManager.getBufferSize
-    )
+    if (log.isDebugEnabled) {
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala]: This is duplicated from the log itself- https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/event/Logging.scala#L1378 

Unless it is primarily focused on the building up of pendingAcks and other args?

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailVehicleManager.scala [hunk 4]
```diff
@@ -173,7 +173,7 @@ class RideHailVehicleManager(val rideHailManager: RideHailManager, boundingBox:
   }
 
   def getIdleVehiclesAndFilterOutExluded: mutable.HashMap[Id[Vehicle], RideHailAgentLocation] = {
-    idleRideHailVehicles.filterNot(elem => rideHailManager.doNotUseInAllocation.contains(elem._1))
+    idleRideHailVehicles.filter(elem => !rideHailManager.doNotUseInAllocation.contains(elem._1))
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/ridehail/RideHailVehicleManager.scala]: Is this preference, or does this actually make a difference in performance?

File: src/main/scala/beam/agentsim/agents/ridehail/allocation/RideHailResourceAllocationManager.scala [hunk 5]
```diff
@@ -176,17 +176,18 @@ abstract class RideHailResourceAllocationManager(private val rideHailManager: Ri
...
-    }.toList
+    }.toVector
     VehicleAllocations(allocResponses)
   }
 
   val repositioningManager: RepositioningManager = createRepositioningManager()
   logger.info(s"Using ${repositioningManager.getClass.getSimpleName} as RepositioningManager")
 
-  def findDepotsForVehiclesInNeedOfRefueling(cavOnly: Boolean = true): Vector[(Id[Vehicle], ParkingStall)] = {
-    val idleVehicleIdsAndLocation: Vector[(Id[Vehicle], RideHailAgentLocation)] =
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/ridehail/allocation/RideHailResourceAllocationManager.scala]: This seems dangerous to push back to the caller. If the filtering of excluded is required then it should be handled inside of the implementation itself.

File: src/main/scala/beam/agentsim/agents/ridehail/repositioningmanager/DemandFollowingRepositioningManager.scala [hunk 6]
```diff
@@ -81,8 +81,11 @@ class DemandFollowingRepositioningManager(val beamServices: BeamServices, val ri
     createClusters
   }
 
-  def repositionVehicles(tick: Int): Vector[(Id[Vehicle], Location)] = {
-    val nonRepositioningIdleVehicles = rideHailManager.vehicleManager.getIdleVehiclesAndFilterOutExluded.values
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/ridehail/repositioningmanager/DemandFollowingRepositioningManager.scala]: Same as above - we shouldn't push this to the calling method - unless it is NOT an implementation detail.

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailVehicleManager.scala (hunk 4 above)
Comment (REASY) [reply src/main/scala/beam/agentsim/agents/ridehail/RideHailVehicleManager.scala]: Yes, it is, here is what I saw in the profiler before optimization (overall)
image

Only RHM thread (before optimization):
image

Only RHM thread (after optimization):
image

File: src/main/scala/beam/agentsim/agents/ridehail/allocation/RideHailResourceAllocationManager.scala (hunk 5 above)
Comment (REASY) [reply src/main/scala/beam/agentsim/agents/ridehail/allocation/RideHailResourceAllocationManager.scala]: @JustinPihony not sure why this is dangerous? I've changed the contract - now you have to provide `idleVehicles`. Maybe I should rename it just to `vehicles`?

File: src/main/scala/beam/agentsim/agents/ridehail/repositioningmanager/DemandFollowingRepositioningManager.scala (hunk 6 above)
Comment (REASY) [reply src/main/scala/beam/agentsim/agents/ridehail/repositioningmanager/DemandFollowingRepositioningManager.scala]: We're repositioning only idle vehicles so it seems ok for me to pass the vehicles to this method.

**Comment by REASY** [comment]
test!

File: src/main/scala/beam/agentsim/agents/ridehail/allocation/RideHailResourceAllocationManager.scala (hunk 5 above)
Comment (REASY) [reply src/main/scala/beam/agentsim/agents/ridehail/allocation/RideHailResourceAllocationManager.scala]: @colinsheppard what do you think?

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala (hunk 3 above)
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala]: I figured

File: src/main/scala/beam/agentsim/agents/ridehail/allocation/RideHailResourceAllocationManager.scala (hunk 5 above)
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/ridehail/allocation/RideHailResourceAllocationManager.scala]: Yah....if the contract is fine to change then I'm ok with that - but it is *definitely* a contract change - if we are ok with the potential for non-idle and exclusionary coming in in other means then that's fine. But if it should always be idle excluded then it is not ok.

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailVehicleManager.scala (hunk 4 above)
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/ridehail/RideHailVehicleManager.scala]: Odd...I'd have to dig in - but will defer to the benchmark. I wonder if the new collections API will make this moot

**Pull request merged**